app/helpers/sessions_helper.rb

```diff
@@ -1,25 +1,51 @@
 module SessionsHelper
 
-    # Logs in the given user.
-    def log_in(user)
-      session[:user_id] = user.id
-    end
-  
-    # Returns the current logged-in user (if any).
-    def current_user
-      if session[:user_id]
-        @current_user ||= User.find_by(id: session[:user_id])
+  # Logs in the given user.
+  def log_in(user)
+    session[:user_id] = user.id
+    # Guard against session replay attacks.
+    session[:session_token] = user.session_token
+  end
+
+  # Remembers a user in a persistent session.
+  def remember(user)
+    user.remember
+    cookies.permanent.encrypted[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
+
+  # Returns the user corresponding to the remember token cookie.
+  def current_user
+    if (user_id = session[:user_id])
+      user = User.find_by(id: user_id)
+      if user && session[:session_token] == user.session_token
+        @current_user = user
+      end
+    elsif (user_id = cookies.encrypted[:user_id])
```

Comment:
So sánh `cookies.encrypted` vs `cookies`?

Reply:
+ cookie : hash value trao đổi giữa client với server thông thường 
+ cookie.encrypted : là hàm có có trong rails nó tự động mã hóa và giải mã --> tăng tính bảo mật

Reply:
cookies.encrypted sẽ mã hóa thông tin của cookie trước khi được đưa vào browser.

Rails sẽ có các cơ chế sau để lưu session:

CookieStore: lưu toàn bộ data trong 1 cookie

CacheStore: lưu data trong cache của Rails

ActiveRecordStore: lưu session data trong database

MemCacheStore: sử dụng Memcached cluster của các serversMặc định là CookieStore